### ISSUE TITLE
Shadowing enum generates NoReturn

### ISSUE THREAD

**Issue by AleCandido** [microsoft/pyright#4409]
**Describe the bug**
It is possible to generate a false `NoReturn` and then false positive for "Code is unreachable" by shadowing an enum variant with its value.

I know shadowing it is not elegant, but it is valid Python code. Indeed, if I annotate the shadowing assignment:
```py
arg: str = arg.value
```
pyright generates an error:
```
Parameter declaration "arg" is obscured by a declaration of the same name
(Pyright reportGeneralTypeIssues)
─────────────────────────────────────────────────────────────────────────
https://github.com/microsoft/pyright/blob/main/docs/configuration.
md#reportGeneralTypeIssues
```

Not sure whether I really need an `enum`, but this is sufficient to generate the error.

**To Reproduce**
Run the self-contained script below.

**Expected behavior**
The function should not have `NoReturn` type, since it is only possible that it raises, but it might regularly

**Screenshots or Code**
```py
import enum

class Greetings(enum.Enum):
    ciao = "come va?" # : Greetings
    hello = "how are you?" # : Greetings

def ciao(arg: Greetings): # -> NoReturn
    arg = arg.value
    if arg not in ["come va?"]:
        raise ValueError("male...")
    return arg

if __name__ == "__main__":
    c = ciao(Greetings.ciao) # : NoReturn
    assert True
    print(c.arg)
```

**VS Code extension or command-line**
I'm using on Neovim through coc https://github.com/fannheyward/coc-pyright, so through the NPM package.

**Comment by erictraut** [comment]
The behavior is correct from a type checking perspective.

The type of `arg` is declared to be `Greetings`. You are assigning an `Any` value to it, which is not a type violation, and the type of `arg` remains `Greetings` after the assignment. The statement `arg not in ["come va?"]` evaluates to True always when `arg` is type `Greeting`.

You have a couple of options:
1. Don't shadow `arg`.
2. Explicitly declare the return type of `ciao` so pyright doesn't attempt to infer it.

**Comment by AleCandido** [comment]
> 1. Don't shadow `arg`.

I know this is an option, and indeed it is what I have done in the original code.

But then are you telling that variables can't change type in a scope? It would be sensible, but it is not what Python is doing.
In particular, if I try to force the assignment with a function:
```py
def come(greet: Greetings) -> str:
    return greet.value

def ciao(arg: Greetings): 
    arg = come(arg.value)
    ...
```
the behavior is still the same, simply now I get a further error from Pyright:
```
Expression of type "str" cannot be assigned to declared type "Greetings"
  "str" is incompatible with "Greetings" (Pyright reportGeneralTypeIssues)
──────────────────────────────────────────────────────────────────────────────
https://github.com/microsoft/pyright/blob/main/docs/configuration.
md#reportGeneralTypeIssues
```

So the problem is not limited to an `Any` assignment.

**Comment by erictraut** [comment]
> But then are you telling that variables can't change type in a scope?

If a variable has a type declaration, that declared type is enforced for all assignments to that variable. If a variable has no type declaration, then pyright allows you to assign any value to it. For more details, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md).

**Comment by AleCandido** [comment]
Ok, thanks for the reference.

I agree that it corresponds to what [is declared](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#type-assignability), but then shadowing is forbidden.

This is not the behavior declared by Python: https://docs.python.org/3/reference/simple_stmts.html#grammar-token-python-grammar-assignment_stmt

It is even compatible with other type systems, e.g. also Rust is type-inferred, but shadowing [is allowed](https://doc.rust-lang.org/reference/statements.html#let-statements):
> Any variables introduced by a variable declaration are visible from the point of declaration until the end of the enclosing block scope, except when they are shadowed by another variable declaration

I know that Python has no `let` statement, but then every assignment should be considered as a `let` statement, otherwise it seems to me that Pyright would be checking its own flavor of Python.

**Comment by erictraut** [comment]
You are confusing two different meanings of the word "shadowing". It generally refers to the case where a symbol in an inner scope has the same name as a symbol in an outer scope. In such a case, there are two distinct symbols, but only the inner-scoped one can be accessed within the inner scope. In cases where there are multiple symbols, each can have its own distinct type declaration. Pyright fully supports this.

```python
x: int = 3

def func1():
    x: str = "" # This is a different symbol from the global `x`

def func2():
    global x
    x = "" # Type error because this `x` refers to the global one
```

The part of the Python spec that you quoted above is talking about this form of shadowing.

Python, unlike Rust, does not generally introduce new scopes for blocks like `if` and `for` statements. Python does introduce a new scope for a list comprehension, classes, lambdas, and functions.

The second type of "shadowing" involves a single symbol (in your first example above, a local variable called `args`) that is assigned values of different types. You can certainly do this in Python. However, if you statically declare the type of that symbol, pyright will enforce that declaration. If you attempt to assign a value to the symbol that violates that declared type, pyright will complain. That's the primary job of a static type checker.

As mentioned in the documentation, if you don't want to declare a type for a symbol, pyright will handle that case as well and allow you to assign any value to the symbol. But in your original example, you declared that the type of `args` was `Greetings`, so pyright will enforce that.

If you are not interested in static type checks, you can disable all type checks by setting the `typeCheckingMode` to "off".

**Comment by AleCandido** [comment]
I was aware of the two meanings. That's why I cited Rust, because the `let` keyword allowed to solve the ambiguity.

I'm sorry for the confusion about the cited docs, but the relevant sentence was just the last one:

> The name is rebound if it was already bound.

Citing Rust once more (short of fantasy), this is an example of how static type check is possible together with second-type shadowing at the same time:
https://doc.rust-lang.org/beta/rust-by-example/variable_bindings/scope.html

So, yes, I'd like to keep static type checks, and a warning about shadowing as bad practice I'd be glad to receive.
Still, I would prefer if the type checker would not give up on valid Python code, because it is not compliant with its own rules.

At this point, being clear that this was the designed behavior, I would call this more a *feature request*, rather than a bug report.
